spring-webflux/src/main/java/org/springframework/web/reactive/result/method/annotation/AbstractMessageWriterResultHandler.java

```diff
@@ -86,9 +87,33 @@ protected AbstractMessageWriterResultHandler(List<HttpMessageWriter<?>> messageW
 	}
 
 
-	@SuppressWarnings("unchecked")
+	/**
+	 * Write a given body to the response with {@link HttpMessageWriter}.
+	 * @param body the object to write
+	 * @param bodyParameter the {@link MethodParameter} of the body to write.
+	 * @param exchange the current exchange.
+	 * @return indicates completion or error
+	 * @see #writeBody(Object, MethodParameter, ServerWebExchange, MethodParameter)
+	 */
 	protected Mono<Void> writeBody(@Nullable Object body, MethodParameter bodyParameter, ServerWebExchange exchange) {
+		return this.writeBody(body, bodyParameter, exchange, null);
+	}
+
+	/**
+	 * Write a given body to the response with {@link HttpMessageWriter}.
+	 * @param body the object to write
+	 * @param bodyParameter the {@link MethodParameter} of the body to write.
+	 * @param actualParameter the actual return type of the method that returned the
+	 * value; could be different from {@code bodyParameter} when processing {@code ResponseEntity}
+	 * for example.
+	 * @param exchange the current exchange.
+	 * @return indicates completion or error
+	 * @since 5.0.2
+	 */
+	@SuppressWarnings("unchecked")
+	protected Mono<Void> writeBody(@Nullable Object body, MethodParameter bodyParameter, ServerWebExchange exchange, @Nullable MethodParameter actualParameter) {
```

Comment:
Let's put ServerWebExchange as the last argument, i.e. `(body, bodyParameter, actualParameter, exchange)`.

Reply:
Indeed we have the same issue with `HttpEntity` I will fix that as well with a similar strategy. Notice that `@RequestBody HttpEntity<Foo> foo entity` does not seems to be currently supported, only `HttpEntity<Foo> foo entity`. Likely an `ArgumentResolver` ordering issue, I will create a JIRA issue for that.